grafana/dm-overview.json

```diff
@@ -98,56 +98,56 @@
             {
               "hide": false,
               "refId": "A",
-              "target": "alias(cloudwatch.request_time_buckets.$environment.router.request_time_bucket_0.samplecount:max, '0-0.025s')"
```

Comment:
So in the past we explicitly chose (but we probably didn't do a very good job on documenting this, my bad, sorry) to use max.

This was because when leadbutt calls for a metric, it says get me metrics for the past 10 minutes. Cloudwatch returns a list of 10 timestamps with their related value.
timestamp10: 100
timestamp9: 89
...
timestamp2: 288
timestamp1: 300
where timestamp 1 is the oldest.

Then it calls again a minute later to get the data and gets a similar list but with 1 minute difference in the time period. (compare this list to the one above)
timestamp11: 49
timestamp10: 100
timestamp9: 89
...
timestamp2: 288

We used to graph based on metrics like `cloudwatch.request_time_buckets.$environment.router.request_time_bucket_0.samplecount:avg`. But we would get weird data were the value shown on the graph was half values like 170.5 which isn't a valid number of requests. It turned out that cloudwatch can give you different data for a very recent timestamp because it is still processing it. So if timestamp 1234567890 gives you value 100 when you first get it, a minute later you might get a slightly higher number like 1234567890: 117 because cloudwatch hadn't finished processing the data the first time round. So we would have the following data in HG.
1234567890: 100
1234567890: 117
And then in terms of the hostedgraphite graphing functions, using `avg` on these would give us a value of `108.5` which is incorrect. Similarly, using `sum` would give us `217` which is again wrong. By using `max` we would get the correct value. Unlike we had previously thought, the HG graphing functions are based on multiple data points for the same timestamp rather than doing an operation on a group of data covering multiple timestamps. This was the best solution we could come up with a reasonable amount of time (although retrospectively I think we should instead have looked at how are asking for metrics from cloudwatch).

So yeah, I would advise that we probably don't want to use the HG `sum` function on data coming through.

In terms of using `sum` on the cloudwatch metrics side of things, that could potentially work though so do investigate and see how you get on but you'll definitely want to try and what we get in hostedgraphite with what we are seeing logs (you could try triggering some 500s/429s yourself and confirming the number in the logs match the numbers you are getting out).

Best of luck. Oh and popping a screenshot of grafana in the PR message when you are ready for a final review would also be useful :)

Reply:
It seems weird that Hosted Graphite works on multiple datapoints if you update a metric with the same timestamp. I've had a search around and other sources suggest that Graphite should overwrite the value for a timestamp with the latest data sent to it. I've pinged and email to their support to see if they have any insights.

Was there any reason for sending the last 10 minutes worth of metrics every minute? Other than for redundancy in case of a network error or our leadbutt app going down of something?

Reply:
This is what I got back from Hosted Graphite. It might be worth recording this somewhere in case it crops up again in future. Might stick in in the manual.

> Hosted Graphite does work differently to normal Graphite. Graphite by default uses a last-write-wins setup so that you can overwrite your datapoints. However that also means that you have to pre-aggregate your datapoints to match the resolution of your storage scheme.

> Hosted Graphite always merges datapoints you send us for the same timestamp, in order to create our Data Views and to ensure that users don't need any kind of separate aggregation service like StatsD or carbon-aggregator. That works perfectly for real time metrics, but it does mean that updating historic datapoints with specific values isn't possible."